fix #3018

polc opened the pull request:
Fix #3018. 

Still a WIP, adding desc on the first order_by column doesn't work yet. Is `diesel_compile_tests/` it a good place to add these tests ? 

This code gives me the following error : 

```rust
let _ = users::table
    .distinct_on(users::id)
    .select(users::id)
    .order_by((users::id.desc(), users::name));
```
```
35 |         .order_by((users::id.desc(), users::name));
|          ^^^^^^^^ the trait `query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<columns::id>>` is not implemented for `diesel::query_builder::order_clause::OrderClause<(diesel::expression::operators::Desc<columns::id>, columns::name)>`
|
= help: the following implementations were found:
<diesel::query_builder::order_clause::OrderClause<(T,)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<T>>>
<diesel::query_builder::order_clause::OrderClause<(__D, T0)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<__D>>>
<diesel::query_builder::order_clause::OrderClause<(__D, T0, T1)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<__D>>>
<diesel::query_builder::order_clause::OrderClause<(__D, T0, T1, T2)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<__D>>>
and 14 others
= note: required because of the requirements on the impl of `OrderDsl<(diesel::expression::operators::Desc<columns::id>, columns::name)>` for `SelectStatement<FromClause<users::table>, diesel::query_builder::select_clause::SelectClause<columns::id>, DistinctOnClause<columns::id>>`
```

weiznich commented:
Thanks for opening this pr :+1:

>  Is diesel_compile_tests/ it a good place to add these tests ?

`diesel_compile_tests` is for verifying that expressions are accepted or rejected by the compiler, so is sort of fine to add the tests there. I tend to add tests that compile to `diesel_tests` as we additionally check the functionality there.

>  This code gives me the following error :
> ```rust
> let _ = users::table
>    .distinct_on(users::id)
>    .select(users::id)
>    .order_by((users::id.desc(), users::name));
>```
>```
>35 |         .order_by((users::id.desc(), users::name));
>|          ^^^^^^^^ the trait `query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<columns::id>>` is not implemented >for `diesel::query_builder::order_clause::OrderClause<(diesel::expression::operators::Desc<columns::id>, columns::name)>`
>|
>= help: the following implementations were found:
><diesel::query_builder::order_clause::OrderClause<(T,)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<T>>>
><diesel::query_builder::order_clause::OrderClause<(__D, T0)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<__D>>>
><diesel::query_builder::order_clause::OrderClause<(__D, T0, T1)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<__D>>>
><diesel::query_builder::order_clause::OrderClause<(__D, T0, T1, T2)> as query_dsl::order_dsl::ValidOrderingForDistinct<DistinctOnClause<__D>>>
>and 14 others
>= note: required because of the requirements on the impl of `OrderDsl<(diesel::expression::operators::Desc<columns::id>, columns::name)>` for `SelectStatement<FromClause<users::table>, diesel::query_builder::select_clause::SelectClause<columns::id>, DistinctOnClause<columns::id>>`
>```

That's because the impl here: https://github.com/diesel-rs/diesel/blob/45823853ed795cbe5a78a8a7d7d0f7a4cd7883fd/diesel/src/type_impls/tuples.rs#L373

Only covers cases where the types match. Fixing that does likely require quite a bit of work, so at least I would be fine with ignoring those cases for now.

polc commented:
I've added a test to `diesel_tests` and removed the sort part so I think it's good for review now. 

Should I create an issue about the sort of the first column for later reference ?

weiznich commented:
> Should I create an issue about the sort of the first column for later reference ?

That would be great

polc commented:
thanks you for the guidance and the fast follow-up :)
